### PR TITLE
[Bootstrapper] Fetch information from dotnet feeds

### PR DESCRIPTION
**Pull request by edvilme**
dotnet/sdk#50601 

Fetch information from dotnet feeds for search and install command. 
Added logic to find global.json file and fetch version information. 
Added logic for downloading files. 

## Search
Added search command to list all available sdk installs.
```
bootstrapper search [<channel>] [--allow-previews]
```
If no channel is provided, it will list all of them by default. 

## Install 
Added install command to install all assets from a specified version. 
```
bootstrapper install [<version>] [--allow-previews]
```
If no version is provided, it will look up the nearest global.json file and install the version specified there.

## Usage example

image

### CODE REVIEW

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        Console.WriteLine($"Installing {product.ProductName} {latestRelease.Version}...");
+
+        foreach (ReleaseComponent component in latestRelease.Components)
+        {
+            Console.WriteLine($"Installing {component.Name} {component.DisplayVersion}...");
+
+            ReleaseFile releaseFile = component.Files.FirstOrDefault(file => file.Rid.Equals(Environment.OSVersion.Platform.ToString(), StringComparison.OrdinalIgnoreCase));
+            
+            releaseFile.DownloadAsync(Path.Combine(path, ".net", component.Name)).Wait();
```
Comment (edvilme) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:49]: TODO: Logic for figuring out where to install to

File: src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+namespace Microsoft.DotNet.Tools.Bootstrapper.Commands.Search;
+
+internal class SearchCommand(
+    ParseResult parseResult) : CommandBase(parseResult)
+{
+    private string _channel = parseResult.ValueForArgument(SearchCommandParser.ChannelArgument);
+    private bool _allowPreviews = parseResult.ValueForOption(SearchCommandParser.AllowPreviews);
+    public override int Execute()
+    {
+        List<Product> productCollection = [.. ProductCollection.GetAsync().Result];
```
Comment (joeloff) [inline src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs:19]: The releases.json only changes about once a month. You should consider caching the file on disk and only update it if there's a new version available.

Comment (baronfel) [reply src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs:19]: We should be using the same etag-based cache invalidation system that @nagilson has for the VSCode extension. ETags are _the_ way to handle cache invalidation of HTTP-delivered resources, and we shouldn't be reinventing the wheel.

File: src/dotnet-bootstrapper/BootstrapperUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                throw new FileNotFoundException("No global.json file found in the directory tree.");
+            }
+            string directoryPath = Path.GetDirectoryName(globalJsonPath);
+            if (directoryPath == null)
+            {
+                throw new DirectoryNotFoundException("Directory path is null.");
+            }
+
+            // TODO: Replace with the actual installation directory.
+            return Path.Combine(directoryPath, ".dotnet.local");
```
Comment (edvilme) [inline src/dotnet-bootstrapper/BootstrapperUtilities.cs:69]: Where should we install to? Should we install to the `.dotnet` directory?

File: src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs [hunk 4]
```diff
@@ -43,10 +43,15 @@ public override int Execute()
...
             foreach (ProductRelease release in releases)
             {
+                // Get release.Sdks latest version
+                var latestSdk = release.Sdks
+                    .OrderByDescending(sdk => sdk.Version)
+                    .FirstOrDefault();
+
                 productMetadataTable.AddRow(
                     release.Version.ToString(), 
                     release.ReleaseDate.ToString("yyyy-MM-dd"),
```
Comment (joeloff) [inline src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs:53]: This forces the US approach for writing dates. Output, even from the CLI should be considered as a UI I think and use the UI thread culture for date formatting. This would get flagged under globalization during 1CS. Consider using the shortformat ("d") specifier. You should be able to do something like `ToString("d", CultureInfo.CurrentUICulture)`

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        return 0;
+    }
+
+    private static void DownloadAndExtractReleaseComponentFiles(ReleaseComponent component, string basePath)
+    {
+        if (component is WindowsDesktopReleaseComponent && !OperatingSystem.IsWindows())
+        {
+            return;
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:68]: Good call and edge case detection here 👌

Comment (nagilson) [reply src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:68]: Probably also worth writing a test for.

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        ReleaseFile releaseFile = component.Files.FirstOrDefault(file =>
+            file.Rid.Equals(BootstrapperUtilities.GetRID(), StringComparison.OrdinalIgnoreCase) && (file.Name.EndsWith(".zip") || file.Name.EndsWith(".tar.gz")));
+
+        if (string.IsNullOrEmpty(releaseFile?.FileName))
+        {
+            Console.WriteLine($"\tNo suitable file found for {component.Name}");
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:76]: Nit: might be helpful to dump the component files array in the error.

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    private static void DownloadAndExtractReleaseComponentFiles(ReleaseComponent component, string basePath)
+    {
+        if (component is WindowsDesktopReleaseComponent && !OperatingSystem.IsWindows())
+        {
+            return;
+        }
+
+        ReleaseFile releaseFile = component.Files.FirstOrDefault(file =>
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:71]:
```suggestion
        ReleaseFile releaseArchiveFile = component.Files.FirstOrDefault(file =>
```

nit: naming of the variable could specify that it is a file that represents just the component that has the zip or tar.gz, instead of any part of the release file. If I understand the code correctly 🤔

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            file.Rid.Equals(BootstrapperUtilities.GetRID(), StringComparison.OrdinalIgnoreCase) && (file.Name.EndsWith(".zip") || file.Name.EndsWith(".tar.gz")));
+
+        if (string.IsNullOrEmpty(releaseFile?.FileName))
+        {
+            Console.WriteLine($"\tNo suitable file found for {component.Name}");
+            return;
+        }
+
+        string zipPath = Path.Combine(basePath, releaseFile.FileName);
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:80]:
```suggestion
        string archiveInstallPath = Path.Combine(basePath, releaseFile.FileName);
```
Nit: Update variable name to be os-agnostic and indicate that it is something that does not yet exist.

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+        if (File.Exists(zipPath))
+        {
+            Console.WriteLine($"\t{component.Name} already exists at {zipPath}");
+            return;
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:85]: I don't think we should return if the zip already exists, as this could allow someone to get into a 'permabad' state. Instead, we should skip the download of the zip but still extract it, as long as it looks like a valid zip (maybe check the sha / file size)

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        try
+        {
+            releaseFile.DownloadAsync(zipPath)?.Wait();
+
+            // Extract the downloaded file
+            ZipFile.ExtractToDirectory(zipPath, Path.ChangeExtension(zipPath, ""));
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:93]: It would be ideal to check whether it has already been extracted to this directory before so we can skip this labor, but iff the extraction resulted in a success and valid installation.

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+            Console.WriteLine($"\tExtracted {component.Name} to {Path.ChangeExtension(zipPath, "")}");
+
+            // Delete the downloaded file
+            File.Delete(zipPath);
```
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:98]: We should probably have an issue w.r.t caching this zip file in case someone installs again? Not very high priority and definitely not a blocker.

File: src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs (hunk 10 above)
Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:93]: I see that DownloadAsync verifies the hash. It's a bit awkward though in that it will do a download even if the dest file exists and always replace it. Ideally we could verify the hash of an existing file and re-utilize it, though I know that's not as relevant in the current context, since we delete the file every time.

Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:93]: One thing to consider: What if someone runs multiple bootstrapper commands at the same time? Will the directory extraction clobber over each other? Should there be some sort of lock on this operation that can be shared across processes? 

What if someone pkills the bootstrapper as it is in the middle of extracting the directory? Then it will leave the directory in a bad state, I believe. We may want to consider cleaning the directory up beforehand (admittedly, this would be bad if someone incorrectly configured their path and then it wiped something else out) or handling these scenarios better with a different approach.

I also don't know if I would block on this but I do consider it higher priority after this.

Comment (nagilson) [inline src/dotnet-bootstrapper/Commands/Install/InstallCommand.cs:93]: It's likely worth doing a dive into each of the possible error cases here to handle them: PathTooLong (does this function respect the OS setting or rely on the default?) -> Directory Not Found - Do we need to create the directory? Unauthorized -> Should we try a fallback directory (probably not for now)

**Review by nagilson — COMMENTED** [review]
Thank you for writing this up, I'm excited to see a lot of progress here!

In the very early days of a project, it's very hard to say what should/shouldnt go in a separate PR and what the 'bar' should be. Some of what I mention below I do think are things we don't need to block this PR on (I mostly remark when that is the case), though some of them are. I think this is a great start 🔥 and it is normal to have a lot of feedback and things to consider.

There is a balance of 'moving fast' vs carefully considering every possible edge case. Since we are not even a launched product I would say we are on the 'moving fast' side for now, to some extent. I would defer to @baronfel and @marcpopMSFT for a broader priority discussion on some of these items.

File: src/dotnet-bootstrapper/BootstrapperUtilities.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                .ToString();
+
+            ReleaseVersion version = ReleaseVersion.Parse(sdkVersion);
+            Console.WriteLine($"Found version {version.Major}.0 in global.json");
+            return $"{version.Major}.0";
+        }
+
+        public static string GetInstallationDirectoryPath()
+        {
+            string globalJsonPath = GlobalJsonUtilities.GetNearestGlobalJsonPath(Environment.CurrentDirectory);
```
[the remark this answers is not in the document]
Comment (nagilson) [reply src/dotnet-bootstrapper/BootstrapperUtilities.cs:57]: I think we can defer this to https://github.com/dotnet/cli-lab/issues/354.

File: src/dotnet-bootstrapper/BootstrapperUtilities.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            if (globalJsonPath == null)
+            {
+                throw new FileNotFoundException("No global.json file found in the directory tree.");
```
[the remark this answers is not in the document]
Comment (nagilson) [reply src/dotnet-bootstrapper/BootstrapperUtilities.cs:60]: https://github.com/dotnet/cli-lab/issues/354

File: src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs (hunk 2 above)
Comment (edvilme) [reply src/dotnet-bootstrapper/Commands/Search/SearchCommand.cs:20]: From what I've seen, I don't think that it does. I do agree that it would make a lot of sense to implement it there

File: src/dotnet-bootstrapper/GlobalJsonUtilities.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.IO;
+using System.Text.Json;
+
+namespace Microsoft.DotNet.Tools.Bootstrapper
+{
+    internal static class GlobalJsonUtilities
+    {
+        public static string GetNearestGlobalJsonPath(string basePath)
```
[the remark this answers is not in the document]
Comment (nagilson) [reply src/dotnet-bootstrapper/GlobalJsonUtilities.cs:13]: > This is good for now, I think the runtime implemented some global json logic and it loks like CliTemplateEngineHost also did. We should probably try to share that implementation in the future, and I can create an issue for that.

https://github.com/dotnet/cli-lab/issues/393

**Comment by nagilson** [comment]
Great work updating things in response to my feedback above; if there's anything I can clarify then please let me know, happy to chat! 😊

**Comment by marcpopMSFT** [comment]
Note: Long term effort that needs reevaluation for our backlog.

**Comment by lextm** [comment]
Fully understand that the usefulness of metadata from .NET feeds but maybe dotnet-core-uninstall should remain as simple as possible (no network connectivity required).

In case end users do need to see more information about the versions they installed, a wrapper tool can combine both the metadata and dotnet-core-uninstall output and provide a streamlined view. [DotUninstall](https://github.com/lextudio/DotUninstall) is an example.
